lib/SILOptimizer/Mandatory/MandatoryCombine.cpp

```diff
@@ -0,0 +1,222 @@
+#endif
+      LLVM_DEBUG(llvm::raw_string_ostream SS(instructionDescription);
+                 instruction->print(SS); instructionDescription = SS.str(););
+      LLVM_DEBUG(llvm::dbgs()
+                 << "MC: Visiting: " << instructionDescription << '\n');
+
+      if (auto replacement = visit(instruction)) {
+        worklist.replaceInstructionWithInstruction(instruction, replacement
+#ifndef NDEBUG
+            , instructionDescription
+#endif
+          );
+      }
+
+      for (SILInstruction *instruction : instructionsPendingDeletion) {
+        worklist.eraseInstFromFunction(*instruction);
+      }
+      instructionsPendingDeletion.clear();
+
+      // Our tracking list has been accumulating instructions created by the
+      // SILBuilder during this iteration. Go through the tracking list and add
+      // its contents to the worklist and then clear said list in preparation
+      // for the next iteration.
+      for (SILInstruction *instruction : createdInstructions) {
+        LLVM_DEBUG(llvm::dbgs() << "MC: add " << *instruction
+                                << " from tracking list to worklist\n");
+        worklist.add(instruction);
+      }
+      createdInstructions.clear();
+    }
+
+    worklist.resetChecked();
+    return madeChange;
+  }
+
+  void clear() {
+    iteration = 0;
+    worklist.resetChecked();
+    madeChange = false;
+  }
+
+
+  /// Applies the MandatoryCombiner to the provided function.
+  ///
+  /// \param function the function to which to apply the MandatoryCombiner.
+  ///
+  /// \return whether a change was made.
+  bool runOnFunction(SILFunction &function) {
+    bool changed = false;
+
+    while (doOneIteration(function, iteration)) {
+      changed = true;
+      ++iteration;
+    }
+
+    return changed;
+  }
+};
+
+} // end anonymous namespace
+
+//===----------------------------------------------------------------------===//
+//                            Top Level Entrypoint
+//===----------------------------------------------------------------------===//
+
+namespace {
+
+class MandatoryCombine final : public SILFunctionTransform {
+
+  llvm::SmallVector<SILInstruction *, 64> createdInstructions;
+
+  void run() override {
+    auto *function = getFunction();
+
+    MandatoryCombiner combiner(createdInstructions);
+    bool madeChange = combiner.runOnFunction(*function);
+
+    if (madeChange) {
+      invalidateAnalysis(SILAnalysis::InvalidationKind::Instructions);
+    }
+  }
+
+  void handleDeleteNotification(SILNode *node) override {
+    // Remove instructions that were both created and deleted from the list of
+    // created instructions which will eventually be added to the worklist.
+
+    auto instruction = dyn_cast<SILInstruction>(node);
+    if (instruction == nullptr) {
+      return;
+    }
+
+    // Linear searching the tracking list doesn't hurt because usually it only
+    // contains a few elements.
+    auto iterator = std::find(createdInstructions.begin(),
```

Comment:
There is a range adaptor for this in include/swift/Basic/STLExtras. The nice thing about the range adapters is they make the code more terse since you only write one parameter instead of two.

In fact, check this:

```
if (createdInstructions.end() != find(createdInstructions, instruction)) {
  createdInstructions.erase(iterator);
}
```

Reply:
Ah, cool

Where is iterator defined here?  Would this need to be

```
auto iterator = find(createdInstructions, instruction);
if (createdInstructions.end() != iterator) {
    createdInstructions.erase(iterator);
}
```